lib/jobs/generate-tag.js

```diff
@@ -66,36 +71,53 @@ function generateTagJobFactory(
                 })
                 .uniq()
                 .value();
-            return [ Promise.map(catalogTypes, function(type) {
+
+            if (self.nodes) {
+                return Promise.map(self.nodes, function(nodeId) {
+                    return Promise.map(catalogTypes, function(type) {
+                        return waterline.catalogs.findMostRecent({
+                            node: nodeId,
+                            source: type
+                        });
+                    }).then(processCatalogs.bind(null, tags, nodeId));
+                });
+            }
+
+            return Promise.map(catalogTypes, function(type) {
                 return waterline.catalogs.findMostRecent({
                     node: self.nodeId,
                     source: type
                 });
-            }), tags ];
-        }).spread(function(catalogs, tags) {
-            catalogs = _(catalogs)
-                .flattenDeep()
-                .compact()
-                .transform(function (catalogs, catalog) {
-                    catalogs[catalog.source] = catalog.data;
-                }, {})
-                .value();
-
-            var matches = _.filter( matchTags(catalogs, tags), function(match) {
-                return match.tag.rules.length && !match.errors.length;
-            });
+            }).then(processCatalogs.bind(null, tags, self.nodeId));
 
-            var tagNames = _(matches).pluck('tag.name').value();
-            if( !_.isEmpty(tagNames)) {
-                return waterline.nodes.addTags(self.nodeId, tagNames);
-            }
         }).then(function() {
             self._done();
         }).catch(function(err) {
             self._done(err);
         });
     };
 
+    function processCatalogs(tags, nodeId, catalogs) {
+        catalogs = _(catalogs)
+            .flattenDeep()
+            .compact()
+            .transform(function (catalogs, catalog) {
+                catalogs[catalog.source] = catalog.data;
+            }, {})
+            .value();
+
+        var matches = _.filter( matchTags(catalogs, tags), function(match) {
+            return match.tag.rules.length && !match.errors.length;
```

Comment:
this is not the out-most return, just within the callback of `filter`, so it doesn't need be promise.